src/erl.rs

```diff
@@ -142,30 +143,108 @@ pub fn record_definition(name: &str, fields: &[&str]) -> String {
+                opaque,
+                ..
+            } => {
+                // Type Exports
+                if *public {
+                    type_exports.push(
+                        name.to_snake_case()
+                            .to_doc()
+                            .append("/")
+                            .append(parameters.len()),
+                    )
+                }
+                // Type definitions
+                let constructors = concat(
+                    constructors
+                        .iter()
+                        .map(|c| {
+                            // TODO more constructor stuff
+                            // {name, values, values, ...}
+                            let name = c.name.to_snake_case().to_doc();
+                            let constructor = if c.args.is_empty() {
+                                name
+                            } else {
+                                name.append(", ").append(concat(
+                                    c.args
+                                        .iter()
+                                        .map(|a| a.1.to_erlang_type_def())
+                                        .intersperse(", ".to_doc()),
+                                ))
+                            };
+                            "{".to_doc().append(constructor).append("}")
+                        })
+                        .intersperse(" | ".to_doc()),
+                )
+                .append(".");
+                let params = concat(
+                    parameters
+                        .iter()
+                        .map(|p| variable_name(p).to_doc())
+                        .intersperse(", ".to_doc()),
+                );
+                let doc = if *opaque { "-opaque " } else { "-type " }
+                    .to_doc()
+                    .append(name.to_snake_case())
+                    .append("(")
+                    .append(params)
+                    .append(") :: ")
+                    .append(constructors);
+                type_defs.push(doc);
+            }
 
-                Statement::ExternalFn {
-                    public: true,
-                    name,
-                    args,
-                    ..
-                } => Some((name.clone(), args.len())),
+            _ => {}
+        }
+    }
 
-                _ => None,
-            })
-            .map(|(n, a)| atom(n).append("/").append(a))
-            .intersperse(", ".to_doc()),
-    );
+    let exports = if exports.is_empty() {
+        nil()
+    } else {
+        "-export(["
+            .to_doc()
+            .append(concat(exports.into_iter().intersperse(", ".to_doc())))
+            .append("]).")
+            .append(lines(2))
+    };
+
+    let type_exports = if type_exports.is_empty() {
+        nil()
+    } else {
+        "-export_type(["
+            .to_doc()
+            .append(concat(type_exports.into_iter().intersperse(", ".to_doc())))
```

Comment:
Let's wrap these with a flex break

https://github.com/gleam-lang/gleam/blob/eb669e955be6b930379e236fe0f1872ae7844c39/src/format.rs#L240-L250

It would be nice to also do this for the regular exports but I understand if you would rather not add that to this PR.

Reply:
happy to, not entirely sure what I'm doing here. :) I'll have to play around with the `pretty::` stuff, it's fairly abstract to me atm.

Reply:
It is super abstract.

The basic idea is that any place there is a `break_` the printer may decide to render a newline after the contained string if the line is too long. It uses one string for when rending on 2 lines, the other for when on 1 line.

When a line is too long all breaks within a `group` render as a newline. Bit hard to explain but it can be good to add a group to each thing you would split by itself when manually formatting. i.e. put a group around each list, each function call, etc.

Reply:
If you're up for a little reading the paper linked from the pretty module is pretty readable as far as papers go

Reply:
I think I mostly got it. Great paper and interesting algorithm / data structure. Got carried away writing some code to better understand it, https://hexdocs.pm/pretty_print/pretty_print/ 🦩

Reply:
Ooh very nice!